infra/src/main/kotlin/fi/fta/geoviite/infra/tracklayout/LayoutContextData.kt

```diff
@@ -0,0 +1,280 @@
+import com.fasterxml.jackson.annotation.JsonIgnore
+import fi.fta.geoviite.infra.common.*
+import fi.fta.geoviite.infra.common.DataType.STORED
+import fi.fta.geoviite.infra.common.DataType.TEMP
+import fi.fta.geoviite.infra.logging.Loggable
+
+// TODO: GVT-2426 This enum is an outdated concept that won't serve well with designs. It's only used by the front. Can we refactor it out?
+enum class ContextType {
+    OFFICIAL, NEW_DRAFT, EDITED_DRAFT, DESIGN, NEW_DESIGN_DRAFT, EDITED_DESIGN_DRAFT
+}
+
+interface LayoutContextAware<T> {
+    val id: DomainId<T>
+    val dataType: DataType
+    // TODO: GVT-2426 if we keep this, the field should be renamed
+//    val contextType: ContextType
+    val draftType: ContextType
+    @get:JsonIgnore val rowId: DomainId<T>
+    @get:JsonIgnore val isDraft: Boolean
+    @get:JsonIgnore val isOfficial: Boolean
+    @get:JsonIgnore val isDesign: Boolean
+    @get:JsonIgnore val isEditedDraft: Boolean
+    @get:JsonIgnore val isNewDraft: Boolean
+}
+
+sealed class LayoutConcept<T : LayoutConcept<T>>(contextData: LayoutContextData<T>) :
+    LayoutContextAware<T> by contextData, Loggable {
+
+    abstract val version: RowVersion<T>?
+    abstract val contextData: LayoutContextData<T>
+}
+
+sealed class LayoutContextData<T> : LayoutContextAware<T> {
+    @get:JsonIgnore
+    override val isDraft: Boolean get() = this is MainDraftContextData || this is DesignDraftContextData
+
+    @get:JsonIgnore
+    override val isOfficial: Boolean get() = this is MainOfficialContextData
+
+    @get:JsonIgnore
+    override val isDesign: Boolean get() = this is DesignContextData
+
+    @get:JsonIgnore
+    override val isEditedDraft: Boolean
+        get() = when (this) {
+            is MainDraftContextData -> officialRowId != null
+            is DesignDraftContextData -> designRowId != null
+            else -> false
+        }
+
+    @get:JsonIgnore
+    override val isNewDraft: Boolean get() = this.isDraft && !isEditedDraft
+
+    @get:JsonIgnore
+    open val officialRowId: DomainId<T>? get() = null
+
+    @get:JsonIgnore
+    open val designRowId: DomainId<T>? get() = null
+
+    companion object {
+        fun <T> newDraft(id: DomainId<T> = StringId()) = MainDraftContextData(id, null, null, TEMP)
+        fun <T> newOfficial(id: DomainId<T> = StringId()) = MainOfficialContextData(id, TEMP)
+    }
+}
+
+sealed class MainContextData<T> : LayoutContextData<T>()
+
+sealed class DesignContextData<T> : LayoutContextData<T>() {
+    abstract val designId: IntId<LayoutDesign>
+}
+
+data class MainOfficialContextData<T>(
+    override val rowId: DomainId<T>,
+    override val dataType: DataType,
+) : MainContextData<T>() {
+    override val id: DomainId<T> = rowId
+    override val draftType: ContextType get() = ContextType.OFFICIAL
+
+    fun asMainDraft(): MainDraftContextData<T> = MainDraftContextData(
+        // TODO: GVT-2426 This is the old way of things, but do we actually need to support switching context in temp objects?
```

Comment:
Tämä härveli toistuu kaikissa. Tuotantokoodin kannalta tämän voisi vain poistaa ja assertoida että datatype ei ole koskaan TEMP silloin kun näitä siirtymiä tehdään. Testikoodissa kuitenkin alustetaan paljon kamaa officialina ja sitten tallennetaan se draftina, muuttaen lennossa. Sekin menisi eksplisiittisemmäksi jos data vain alustettaisiin oikeaan kontekstiin suoraan mutta korjattavia testejä on noin 150 -> teen erikseen tämän pullarin jälkeen.